src/persist-client/src/internal/compact.rs

```diff
@@ -254,14 +275,233 @@ impl Compactor {
         // Compaction is cpu intensive, so be polite and spawn it on the
         // CPU heavy runtime.
         let compact_span = debug_span!("compact::consolidate");
-        cpu_heavy_runtime
+        parts_cpu_heavy_runtime
             .spawn_named(
                 || "persist::compact::consolidate",
                 compact_blocking.instrument(compact_span),
             )
             .await?
     }
 
+    /// WIP: grab the first run from each batch, then the second, etc.
+    fn order_runs<T, D>(req: &CompactReq<T>) -> Vec<(&HollowBatch<T>, &[PartialBatchKey])>
+    where
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        let mut all_runs = vec![];
+        for batch in &req.inputs {
+            all_runs.push((batch, batch.runs()));
+        }
+
+        // map our (Batch, [Runs]) to [(Batch, Run), ...] cycling through the input batches
+        let mut finished_iterators = 0;
+        let mut ordered_runs = vec![];
+        loop {
+            for (batch, runs) in &mut all_runs {
+                match runs.next() {
+                    Some(run) => ordered_runs.push((*batch, run)),
+                    None => finished_iterators += 1,
+                }
+            }
+
+            if finished_iterators >= all_runs.len() {
+                break;
+            }
+        }
+
+        ordered_runs
+    }
+
+    // WIP: Compacts runs together, pulling down one part at a time from each. If the runs are
+    //      already sorted, the result should be a single run. Memory is bounded by the # of
+    //      runs passed into this function, consuming at most (runs + 1) * blob_target_size
+    async fn compact_runs<'a, T, D>(
+        cfg: &'a PersistConfig,
+        shard_id: &'a ShardId,
+        desc: &'a Description<T>,
+        runs: &'a mut [(&'a HollowBatch<T>, &'a [PartialBatchKey])],
+        mut batch_parts: BatchParts<T>,
+        blob: Arc<dyn Blob + Send + Sync>,
+        metrics: Arc<Metrics>,
+    ) -> Result<(Vec<PartialBatchKey>, Vec<usize>, usize), anyhow::Error>
+    where
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64 + Send + Sync,
+    {
+        let mut compaction_runs = vec![];
+        let mut compaction_parts = vec![];
+        let mut total_updates = 0;
+
+        let mut heap = BinaryHeap::new();
+        let mut update_buffer: Vec<((Vec<u8>, Vec<u8>), T, D)> = Vec::new();
+        let mut update_buffer_size_bytes = 0;
+        let mut greatest_kv: Option<(Vec<u8>, Vec<u8>)> = None;
+
+        // we'll reference our runs by their index order in the original `runs` vec
+        let mut remaining_updates_by_run = vec![0; runs.len()];
+        let mut runs: Vec<_> = runs
+            .iter()
+            .map(|(batch, parts)| (batch, parts.into_iter()))
+            .collect();
+
+        // populate our heap with the updates from the first part of each run
+        for (index, (batch, parts)) in runs.iter_mut().enumerate() {
```

Comment:
I think we should be able to do something like wrap `(&HollowBatch<T>, &[PartialBatchKey])` in a struct that we can put directly in the heap. then the loop becomes something like

```
while !heap.is_empty() {
    let mut least_run = heap.pop();
    update_buffer.push(least_run.next());
    if !least_run.is_empty() {
        heap.push(least_run);
    }
}
```

IMO if this works out, it would be easier to follow (and be more resilient to future maintenance) than doing stuff based on array indexes

Reply:
hm I think I'm missing how we order the heap in that scenario. in my mind, that requires `least_run` to contain the next element from the run, so that the heap bubbles up the right one to the top. I'm imagining the type would expand out to something like: `BinaryHeap<(((k, v), t, d), HollowBatch, &[PartialBatchKey])>` 

and in making that a requirement, since the runs aren't pure iterators and need to pull down a full part at a time, it seems like they'd need to first buffer their active part's updates in their own Vec first. So then we add another intermediate step and do: `per-run buffer --> heap --> update_buffer` rather than just the latter two steps now

Reply:
I was imagining a newtype wrapper around that stuff that implemented whatever custom logic we needed for PartialOrd/Ord. but it sounds like my idea was half baked, so feel free to ignore this until I can get a moment to really wrap my head around the details (not going to happen today)

Reply:
prototyped this. it didn't reduce complexity as much as I was hoping so I'm happy to leave this as-is

for posterity

```
diff --git a/src/persist-client/src/internal/compact.rs b/src/persist-client/src/internal/compact.rs
index 0262b0e01..10a075d1e 100644
--- a/src/persist-client/src/internal/compact.rs
+++ b/src/persist-client/src/internal/compact.rs
@@ -7,7 +7,7 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::cmp::Reverse;
+use std::cmp::{Ordering, Reverse};
 use std::collections::BinaryHeap;
 use std::fmt::Debug;
 use std::iter::Peekable;
@@ -16,6 +16,7 @@ use std::sync::Arc;
 use std::time::Instant;
 
 use anyhow::anyhow;
+use async_trait::async_trait;
 use differential_dataflow::consolidation::consolidate_updates;
 use differential_dataflow::difference::Semigroup;
 use differential_dataflow::lattice::Lattice;
@@ -391,66 +392,26 @@ impl Compactor {
         let mut compaction_parts_count = 0;
         let mut total_updates = 0;
 
-        let mut sorted_updates = BinaryHeap::new();
         let mut update_buffer: Vec<((Vec<u8>, Vec<u8>), T, D)> = Vec::new();
         let mut update_buffer_size_bytes = 0;
         let mut greatest_kv: Option<(Vec<u8>, Vec<u8>)> = None;
 
-        let mut remaining_updates_by_run = vec![0; runs.len()];
-        let mut runs: Vec<_> = runs
-            .iter()
-            .map(|(batch, parts)| (batch, parts.into_iter()))
-            .collect();
-
-        // populate our heap with the updates from the first part of each run
-        for (index, (batch, parts)) in runs.iter_mut().enumerate() {
-            if let Some(key) = parts.next() {
-                fetch_batch_part(
-                    &shard_id,
-                    blob.as_ref(),
-                    &metrics,
-                    key,
-                    &batch.desc,
-                    |k, v, mut t, d| {
-                        t.advance_by(desc.since().borrow());
-                        let d = D::decode(d);
-                        let k = k.to_vec();
-                        let v = v.to_vec();
-                        // default heap ordering is descending
-                        sorted_updates.push(Reverse((((k, v), t, d), index)));
-                        remaining_updates_by_run[index] += 1;
-                    },
-                )
-                .await?;
+        let mut heap = BinaryHeap::with_capacity(runs.len());
+        for _run in runs {
+            if let Some(entry) =
+                CompactionHeapEntry::new(Box::new(()) as Box<dyn UpdatesIter<T, D>>).await
+            {
+                heap.push(entry);
             }
         }
 
         // repeatedly pull off the least element from our heap, refilling from the originating run
         // if needed. the heap will be exhausted only when all parts from all input runs have been
         // consumed.
-        while let Some(Reverse((((k, v), t, d), index))) = sorted_updates.pop() {
-            remaining_updates_by_run[index] -= 1;
-            if remaining_updates_by_run[index] == 0 {
-                // repopulate from the originating run, if any parts remain
-                let (batch, parts) = &mut runs[index];
-                if let Some(key) = parts.next() {
-                    fetch_batch_part(
-                        &shard_id,
-                        blob.as_ref(),
-                        &metrics,
-                        key,
-                        &batch.desc,
-                        |k, v, mut t, d| {
-                            t.advance_by(desc.since().borrow());
-                            let d = D::decode(d);
-                            let k = k.to_vec();
-                            let v = v.to_vec();
-                            sorted_updates.push(Reverse((((k, v), t, d), index)));
-                            remaining_updates_by_run[index] += 1;
-                        },
-                    )
-                    .await?;
-                }
+        while let Some(entry) = heap.pop() {
+            let (((k, v), t, d), entry) = entry.next().await;
+            if let Some(entry) = entry {
+                heap.push(entry);
             }
 
             // both T and D are Codec64, ergo 8 bytes a piece
@@ -634,6 +595,63 @@ impl<'a, T> Iterator for HollowBatchRunIter<'a, T> {
     }
 }
 
+/// WIP can this just be futures_util::Stream?
+#[async_trait]
+trait UpdatesIter<T, D>: Send + Sync + Debug {
+    async fn next(&mut self) -> Option<((Vec<u8>, Vec<u8>), T, D)>;
+}
+
+#[async_trait]
+impl<T, D> UpdatesIter<T, D> for () {
+    async fn next(&mut self) -> Option<((Vec<u8>, Vec<u8>), T, D)> {
+        todo!("WIP")
+    }
+}
+
+#[derive(Debug)]
+struct CompactionHeapEntry<T, D> {
+    update: ((Vec<u8>, Vec<u8>), T, D),
+    updates: Box<dyn UpdatesIter<T, D>>,
+}
+
+impl<T: Ord + Debug, D: Ord + Debug> PartialEq for CompactionHeapEntry<T, D> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl<T: Ord + Debug, D: Ord + Debug> Eq for CompactionHeapEntry<T, D> {}
+
+impl<T: Ord + Debug, D: Ord + Debug> PartialOrd for CompactionHeapEntry<T, D> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: Ord + Debug, D: Ord + Debug> Ord for CompactionHeapEntry<T, D> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.update.cmp(&other.update).reverse()
+    }
+}
+
+impl<T: Ord + Debug, D: Ord + Debug> CompactionHeapEntry<T, D> {
+    async fn new(mut updates: Box<dyn UpdatesIter<T, D>>) -> Option<Self> {
+        updates
+            .next()
+            .await
+            .map(|update| CompactionHeapEntry { update, updates })
+    }
+
+    async fn next(mut self) -> (((Vec<u8>, Vec<u8>), T, D), Option<Self>) {
+        let update = self.update;
+        let next = self.updates.next().await.map(|update| CompactionHeapEntry {
+            update,
+            updates: self.updates,
+        });
+        (update, next)
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use timely::progress::Antichain;
```

Reply:
hmm also went down this road and so far the net result is shaping up to be more complex than what's there right now